src/pool/srv_pool.c

```diff
@@ -3761,6 +3761,21 @@ ds_pool_update_internal(uuid_t pool_uuid, struct pool_target_id_list *tgts,
 
 	replace_failed_replicas(svc, map);
 
+	if (opc == POOL_ADD_IN) {
+		/*
+		 * If we are setting ranks from UP -> UPIN, schedule a reclaim
+		 * operation to garbage collect any unreachable data moved
+		 * during reintegration/addition
+		 */
+		rc = ds_rebuild_schedule(pool_uuid, map_version, tgts,
+					 RB_OP_RECLAIM);
```

Comment:
Hmm, I thought we already got the object list during reintegration progress. Do we need another scan phase to get all "reclaim" objects?

Reply:
We do have that information, but at the time when it's generated, it's not actionable. We can't delete anything until reintegration fully succeeds and the pool map is updated - otherwise clients will still be reading from the rank that had this data. 

Instead of taking this approach, we could store this list of all unreachable object IDs... somewhere? Then after reintegration/addition we could iterate that list to delete everything. I'm not sure if the current reintegration API gives us enough information to know whether the current rank will no longer need the data after reintegration is complete - we might need to also extend the placement API for this. 

I think we could go this route, but it's going to be significantly more code, and will have more corner cases to deal with (rebuild already has to handle state changes in the middle of an ongoing rebuild operation).

Reply:
Do we have any code to ensure the safe and proper pool target state transition? For example: when a target is under reclaiming, a new failure happens and this reclaiming target could be selected as spare target again, will we finish current reclaiming procedure or abort it before the new rebuild?

Reply:
Created a ticket DAOS-5675 to track this. I'd like to add this as a separate pull request.